pkg/controllers/customeradmin/controllernamespace/namespace/namespace_controller.go

```diff
@@ -0,0 +1,126 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// Add creates a new CustomerAdminReconciler Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &reconcileNamespace{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("customeradmin-namespace-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(&source.Kind{Type: &corev1.Namespace{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &reconcileNamespace{}
+
+// ReconcileNamespace reconciles a CustomerAdminReconciler object
+type reconcileNamespace struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile receives a namespace event and reconciles rolebindings for
+// customer roles
+// Note:
+// The Controller will requeue the Request to be processed again if the
+// returned error is non-nil or Result.Requeue is true, otherwise upon
+// completion it will remove the work from the queue.
+func (r *reconcileNamespace) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	for _, forbiddenName := range []string{"default", "kube", "openshift", "kubernetes"} {
+		if strings.HasPrefix(request.Name, forbiddenName) {
+			return reconcile.Result{}, nil
+		}
+	}
+	ctx := context.Background()
+	// move to appropriate place
+	// If we are namespace we need to reconcile 3 rolebindings:
+	// - customer-admin
+	// - customer-project-admin
+	// - customer-view
+	logrus.Printf("NSC: Reconciling Namespace %s", request.Name)
+
+	// check for existence of namespace
+	ns := &corev1.Namespace{}
+	err := r.client.Get(ctx, request.NamespacedName, ns)
+	if err != nil {
+		// namespace is not found
+		if kerrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// do not reconcile. namespace is being deleted
+	if ns.Status.Phase == corev1.NamespaceTerminating {
+		logrus.Printf("NSC: namespace %v is being deleted", request.Name)
+		return reconcile.Result{}, nil
+	}
+
+	for rbname := range rbs.RoleBindingNames {
+		nn := types.NamespacedName{
+			Namespace: request.Name,
+			Name:      rbname,
+		}
+		nrb := rbs.RoleBindingNames[rbname]
+		nrb.Namespace = request.Name
+		nn.Name = rbname
+		rb := &rbacv1.RoleBinding{}
+		err := r.client.Get(ctx, nn, rb)
```

Comment:
Note: in the future dropping the Get may be a good optimisation as it will halve the number of API calls.  Can just do a Create and ignore AlreadyExists errors.

Reply:
This becomes much more efficient and clean this way.  Updated.